### DISCUSSION TITLE
How to make PointerPressed event work

### DISCUSSION THREAD

**Discussion by holdxen** [AvaloniaUI/Avalonia#10673] [Q&A]
Binding event handler in xaml does't work

**Comment by timunie** [comment]
Events cannot be bound. You need to define them in code behind

In `MyView.axaml`:
```xml
<Button PointerPressed="MyButton_OnPointerPressed" />
```

In `MyView.axaml.cs` (code behind):
```cs
private void MyButton_OnPointerPressed(object? sender, PointerPressedEventArgs e)
{
    DoStuff()
}
```

**Reply by timunie** [reply]
okay, looks like you need to tunnel this event for a Button to work. For Label it works as shown above, but Button seems to eat that event itself. 

So do the following: 
1. In XAML add a Name and remove the event:
```xml
<Button x:Name="MyButton"
		Content="Click Me" />
```

2. In code behind register the tunnel handler: 
```cs
public MainView()
{
    InitializeComponent();

    MyButton.AddHandler(PointerPressedEvent, Button_OnPointerPressed, RoutingStrategies.Tunnel);
}

private void Button_OnPointerPressed(object? sender, PointerPressedEventArgs e)
{
    Debug.WriteLine("Button Pressed");
}
```

**Reply by pyking-wang** [reply]
Is there any logic in this? 
I don't need an AddHandler in Textblock, but PointerPressedEventArgs only works when an AddHandler is needed in ComBox.

**Reply by timunie** [reply]
it is: https://docs.avaloniaui.net/docs/next/concepts/input/routed-events#routing-strategies

As ComboBox will eat up the event, you have to catch it before the control gets it. So you need to "tunnel" it, and this can only be done in code, see: https://docs.avaloniaui.net/docs/next/get-started/wpf/tunnelling-events

**Reply by pyking-wang** [reply]
I made a mistake earlier, it was TextBox; 
I just want to know which controls will eat up the event

**Reply by buzzware** [reply]
This seems crazy to me. 
So
```<Button Click="MyButton_OnClick" />```
works, but 
```<Button PointerPressed="MyButton_OnPointerPressed" />```
does not !?

I read [about bubbling and tunneling](https://docs.avaloniaui.net/docs/next/concepts/input/routed-events#routing-strategies) but it still doesn't make sense that MyButton_OnPointerPressed would not get called.
* "Bubbling: Event handlers on the event source are invoked"
* "Tunneling: Initially, event handlers at the element tree root are invoked. The routed event then travels a route through successive child elements along the route, **towards the node element that is the routed event source (the element that raised the routed event)**."

It sounds like MyButton_OnPointerPressed should get called in either case.

What is going on? This is pretty basic.

**Reply by buzzware** [reply]
```<Button Tapped="MyButton_OnTapped" />``` works just like Click

**Reply by stevemonaco** [reply]
The `PointerPressed` event is handled within `Button` as you can [see here](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Button.cs#L387). By default, this is a bubbling event, so it is never propagated above the `Button` in the hierarchy because it is already handled.

You need to register a tunneling event, so you can get a separate input event besides the bubbling event.
```xml
<Button x:Name="button" Content="Test Button" />
```
```cs
public MainWindow()
{
    InitializeComponent();
    button.AddHandler(PointerPressedEvent, Button_PointerPressed, Avalonia.Interactivity.RoutingStrategies.Tunnel);
}

private void Button_PointerPressed(object? sender, PointerPressedEventArgs e)
{
    button.Content = new TextBlock { Text = "Pressed!" };
}
```

You should probably create a separate discussion if you have further questions.

**Reply by buzzware** [reply]
That may be a technical solution, but it still looks like a DX bug [https://en.wikipedia.org/wiki/Principle_of_least_astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)

**Reply by stevemonaco** [reply]
There is a lot of incongruity in XAML frameworks in general. Fixing it would require a large amount of tooling for analysis (along with lots of annotations in code) and/or breaking changes from how routed events work, among other areas. XAML doesn't hold your hand and expects you to understand its core concepts. That has some downsides, but bubble events exist so you don't need to check subscriptions with every single parent in the logical tree for permission...which would effectively turn it into a tunneling event. The input system is largely a port from the UWP design which doesn't have a `PreviewPointerPressed` event or such for a reason I don't know.

**Reply by robloo** [reply]
@buzzware 

> That may be a technical solution, but it still looks like a DX bug

In your case it seems obvious it should "just work". However, I would think about it a little differently. The pointer pressed is like a raw input event. One that the button needs to handle and then turn into a "Click". Click is what apps should be using in this context.

 * Click -> An event specific to a Button indicating it has been pressed (high level)
 * PointerPressed -> An input event that Button handles and uses to determine it has been clicked (low level)

There are also some technical reasons PointerPressed and others are the way they are. When you start nesting controls the concept of event routing bubbling/tunneling becomes very important. Without it a whole class of scenarios with nested controls just wouldn't work (Nested scrolling for example, etc.).

In the case of a Button, it must mark PointerPressed as handled in case it is within something else that needs to use PointerPressed as well. Think of a contrived example of a button within a button. When you click the inner button you don't also want the outer button to say it has been clicked.

**Reply by buzzware** [reply]
Does it make sense being available in XAML then ? Is it possible to hide it?

**Reply by timunie** [reply]
No it is not possible to hide it from the XAML due to how the inheritance works. For example some one may want to inherit button and change click logic. then the event could be there again.

**Reply by robloo** [reply]
Other use cases and controls, especially custom ones, require PointerPressed as well.

**Reply by stevemonaco** [reply]
Could there be a feature enhancement where you can mark events with an attribute like `[XamlHidden("Event implementation only supports tunneling")]` which is not inherited and the code completion can be made aware? Similar to pseudoclasses. TBH, I would like ways to restrict the API surface (at least on the XAML side) when certain things are unsupported. It could eliminate a lot of footguns, but there's possibly a smarter way such as external configuration classes.

**Reply by robloo** [reply]
Well, PointerPressed is a valid event. It's just handled by Button here so if you subscribe to it you'll never see it. That certainly isn't true for other controls though. I want to use PointerPressed on Border for example, or TextBlock and that must not be hidden. I don't know how you can hide this for some developers and avoid other developers who need it complaining that it is hidden. I think it has to stay as-is and devs just need to be more aware of using the `Click` event in cases like this. That can be better documented.

**Reply by rabbitism** [reply]
And it behaves the same in wpf. It is something need to remember.

**Reply by stevemonaco** [reply]
Oh right, `PointerPressedEvent` is defined only once in `InputElement` and inherited everywhere else. So you wouldn't be able to mark it elsewhere. Which leaves the configuration approach. Something like the following (naively stealing ideas from EF Core):

```cs
public class ButtonXamlConfiguration : IXamlCompletionConfiguration<Button>
{
    public void Configure(XamlCompletionConfigurationBuilder<Button> builder)
    {
        builder.Property(x => x.PointerPressed).IsUnsupported("Event implementation only supports tunneling");
    }
}
```

This would need to be specific to `Button` and not any inheritors. Code completion could completely hide it or put it at the bottom in a special category, ideally with the message and an separate icon/visual indicator. Extending this to allow themes to further restrict the surface would be ideal where certain dependency properties are ignored. That's a huge tracking feature like `CompiledBinding` to know which `ControlTheme` is being used (in the typical case where it doesn't change). Then, of course, there's building the support for this.

Regardless, even if it's documented, people probably aren't going to read and remember it until they hit the exact problem. I think it's important to be more proactive than docs in this particular area as the IDE is a primary exploratory tool for beginners.
